dodiscover/ci/categorical_test.py

```diff
@@ -0,0 +1,200 @@
+                                             :footcite:`cressieread1984`"
+
+    Returns
+    -------
+    CI Test Results: tuple
+        Returns a tuple (chi, p_value, dof). `chi` is the
+        chi-squared test statistic. The `p_value` for the test, i.e. the
+        probability of observing the computed chi-square statistic (or an even
+        higher value), given the null hypothesis that X \u27C2 Y | Zs is True.
+        If boolean = True, returns True if the p_value of the test is greater
+        than `significance_level` else returns False.
+
+    See Also
+    --------
+    scipy.stats.power_divergence
+
+    References
+    ----------
+    .. footbibliography::
+
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> import numpy as np
+    >>> data = pd.DataFrame(np.random.randint(0, 2, size=(50000, 4)), columns=list('ABCD'))
+    >>> data['E'] = data['A'] + data['B'] + data['C']
+    >>> chi_square(X='A', Y='C', Z=[], data=data, boolean=True, significance_level=0.05)
+    True
+    >>> chi_square(X='A', Y='B', Z=['D'], data=data, boolean=True, significance_level=0.05)
+    True
+    >>> chi_square(X='A', Y='B', Z=['D', 'E'], data=data, boolean=True, significance_level=0.05)
+    False
+    """
+
+    # Step 1: Check if the arguments are valid and type conversions.
+    if hasattr(Z, "__iter__"):
+        Z = list(Z)
+    else:
+        raise (f"Z must be an iterable. Got object type: {type(Z)}")
+
+    if (X in Z) or (Y in Z):
+        raise ValueError(f"The variables X or Y can't be in Z. Found {X if X in Z else Y} in Z.")
+
+    # Step 2: Do a simple contingency test if there are no conditional variables.
+    if len(Z) == 0:
+        chi, p_value, dof, expected = stats.chi2_contingency(
+            data.groupby([X, Y]).size().unstack(Y, fill_value=0), lambda_=lambda_
+        )
+
+    # Step 3: If there are conditionals variables, iterate over unique states and do
+    #         the contingency test.
+    else:
+        chi = 0
+        dof = 0
+        for z_state, df in data.groupby(Z):
+            try:
+                c, _, d, _ = stats.chi2_contingency(
+                    df.groupby([X, Y]).size().unstack(Y, fill_value=0), lambda_=lambda_
+                )
+                chi += c
+                dof += d
+            except ValueError:
+                # If one of the values is 0 in the 2x2 table.
+                if isinstance(z_state, str):
+                    logging.info(
+                        f"Skipping the test {X} \u27C2 {Y} | {Z[0]}={z_state}. Not enough samples"
+                    )
+                else:
+                    z_str = ", ".join([f"{var}={state}" for var, state in zip(Z, z_state)])
+                    logging.info(f"Skipping the test {X} \u27C2 {Y} | {z_str}. Not enough samples")
+        p_value = 1 - stats.chi2.cdf(chi, df=dof)
+        import numpy as np
+
+        if np.isnan(p_value):
+            print(p_value, chi, c, dof, X, Y, Z)
```

Comment:
Avoid putting print statements. Maybe instead use a logger as in the lines above.